keras/caffe/converters.py

```diff
@@ -0,0 +1,399 @@
+from ..layers.convolutional import *
+from ..layers.core import *
+from ..layers.normalization import *
+
+from ..models import Graph
+
+import caffe_pb2 as caffe
+import google.protobuf
+from caffe_utils import *
+
+import numpy as np
+import math
+
+def is_data_input(layer):
+	if layer.type == 5 or layer.type == 12 or layer.type == 24 or layer.type == 29 or layer.type == 9:
+		return True
+	else:
+		return False
+
+
+def model_from_config(layers, phase, input_dim):
+	'''
+		layers: a list of all the layers in the model
+		phase: parameter to specify which network to extract: training or test
+		input_dim: input dimensions of the configuration (if in model is in deploy mode)
+	'''
+	# DEPLOY MODE: first layer is computation
+	# NON DEPLOY MODE: first layer is data input
+	if input_dim == []:
+		in_deploy_mode = False
+	else:
+		in_deploy_mode = True
+
+	network = make_network(layers, phase)	# obtain the nodes that make up the graph
+	if len(network) == 0:
+		raise Exception('failed to construct network from the prototext')
+	network = acyclic(network)	# Convert it to be truly acyclic
+	network = merge_layer_blob(network)	# eliminate 'blobs', just have layers
+	reverse_network = reverse(network)	# reverse, to obtain the start
+	starts = get_starts(reverse_network)	# inputs of the network - 'in-order' is zero
+	ends = get_ends(network)	# outputs of the network - 'out-order' is zero
+	network = remove_label_paths(network, starts, ends)	# path from input to loss layers removed
+	reverse_network = reverse(network) # stores the 'input' for each layer
+
+	model = Graph()
+
+	output_dim = {}	# for dimensionality inference (layer_nb: output_dim)
+
+	# add input nodes
+	for start in starts:
+		name = layers[start].name
+		model.add_input(name=name, ndim=4)
+		if in_deploy_mode:
+			output_dim[name] = input_dim
+		else:
+			data_dim = get_data_dim(layers[start])	# not in deploy mode - need to get input DATA dimensions from layers and their transformations
+			output_dim[name] = data_dim
+			
+
+	# parse all the layers and build equivalent keras graph
+	for layer_nb in network:
+		layer = layers[layer_nb]
+		name = layer.name
+
+		if is_data_input(layer):
+			# not including data layers
+			continue
+
+		if layer_nb in starts:
+			# in deploy mode since data layers can't reach this line and condition holds.
+			new_name = 'graph_input_' + name
+			input_layer_names = [name]
+			name = new_name
+		else:
+			input_layers = reverse_network[layer_nb]	# inputs to current layer, in the form of layer numbers
+			input_layer_names = []	# list of strings identifying the input layer
+			for input_layer in input_layers:
+				if input_layer in starts and in_deploy_mode:
+					input_layer_names.append('graph_input_' + layers[input_layer].name)
+				else:
+					input_layer_names.append(layers[input_layer].name)
+
+		if layer_nb in ends:
+			name = 'output_' + name 	# outputs nodes are marked with 'output_' prefix from which output is derived later in 'add_output'
```

Comment:
To avoid very long lines, I would recommend putting comments before the line (possibly over several lines).